ufl/cell.py

```diff
@@ -1,298 +1,456 @@
-"Types for representing a cell."
+"""Types for representing a cell."""
 
 # Copyright (C) 2008-2016 Martin Sandve Alnæs
 #
 # This file is part of UFL (https://www.fenicsproject.org)
 #
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 
+from __future__ import annotations
 import functools
 import numbers
+import typing
 
-import ufl.cell
-from ufl.core.ufl_type import attach_operators_from_hash_data
+from ufl.core.ufl_type import UFLObject
+from abc import abstractmethod
+
+try:
+    from typing import Self
+except ImportError:
+    # Hack to get this to work pre Python 3.11
+    # Delete this after 04 Oct 2026 (Python 3.10 end of life)
+    # After this date, can also replace "self: Self," with "self,
+    from typing import TypeVar
+    Self = TypeVar("Self", bound="AbstractCell")
 
-# Export list for ufl.classes
 __all_classes__ = ["AbstractCell", "Cell", "TensorProductCell"]
 
 
-# --- The most abstract cell class, base class for other cell types
+class AbstractCell(UFLObject):
+    """A base class for all cells."""
+    @abstractmethod
+    def topological_dimension(self) -> int:
```

Comment:
Could we make things properties rather than methods where appropriate?

Reply:
I was planning to do that in a separate PR.

I was aiming for this PR to need no changes (or [negligible changes](FEniCS/ffcx#547)) in FFCx, DOLFINx, TSFC, and Firedrake so we could get it merged more easily and I can start experimenting with whether its useful to implement cells in the Basix UFL wrapper.

Reply:
I've opened #169 and #170 for things I'm planning to do following this that will change the interfaces to elsewhere